File: tests/bool_tests.rs
mod dt_test_utils;
use dt_test_utils::{dt, dt_oneliner};

pub const DT_PATH: &str = std::env!("CARGO_BIN_EXE_dt");

#[test]
fn test_true() {
    let res = dt(&["true", "print"]);
    assert_eq!("", &res.stderr);
    assert_eq!("true", &res.stdout);
    assert!(res.status.success());
}

#[test]
fn test_false() {
    let res = dt(&["false", "print"]);
    assert_eq!("", &res.stderr);
    assert_eq!("false", &res.stdout);
    assert!(res.status.success());
}

#[test]
fn test_not() {
    assert_eq!("true", &dt_oneliner("false not print").stdout);
    assert_eq!("false", &dt_oneliner("true not print").stdout);
}

#[test]
fn test_bool_equality() {
    let first_result = dt_oneliner("true true eq? print");
    assert_eq!("", &first_result.stderr);
    assert_eq!("true", &first_result.stdout);
    assert_eq!("true", &dt_oneliner("false false eq? print").stdout);
    assert_eq!("true", &dt_oneliner("true false neq? print").stdout);
    assert_eq!("true", &dt_oneliner("false true neq? print").stdout);

    assert_eq!("false", &dt_oneliner("true true neq? print").stdout);
    assert_eq!("false", &dt_oneliner("false false neq? print").stdout);
    assert_eq!("false", &dt_oneliner("true false eq? print").stdout);
    assert_eq!("false", &dt_oneliner("false true eq? print").stdout);
}

#[test]
fn test_numeric_equality() {
    assert_eq!("true", &dt_oneliner("1 1 eq? print").stdout);
    assert_eq!("true", &dt_oneliner("1 1.0 eq? print").stdout);
    assert_eq!("true", &dt_oneliner("1.0 1 eq? print").stdout);
    assert_eq!("true", &dt_oneliner("1.0 1.0 eq? print").stdout);

    assert_eq!("false", &dt_oneliner("1 2 eq? print").stdout);
    assert_eq!("false", &dt_oneliner("1 2.0 eq? print").stdout);
    assert_eq!("false", &dt_oneliner("1.0 2 eq? print").stdout);
    assert_eq!("false", &dt_oneliner("1.0 2.0 eq? print").stdout);

    assert_eq!("false", &dt_oneliner("1 1 neq? print").stdout);
    assert_eq!("false", &dt_oneliner("1 1.0 neq? print").stdout);
    assert_eq!("false", &dt_oneliner("1.0 1 neq? print").stdout);
    assert_eq!("false", &dt_oneliner("1.0 1.0 neq? print").stdout);

    assert_eq!("true", &dt_oneliner("1 2 neq? print").stdout);
    assert_eq!("true", &dt_oneliner("1 2.0 neq? print").stdout);
    assert_eq!("true", &dt_oneliner("1.0 2 neq? print").stdout);
    assert_eq!("true", &dt_oneliner("1.0 2.0 neq? print").stdout);
}

#[test]
fn test_string_equality() {
    assert_eq!("true", &dt_oneliner(r#""apple" "apple" eq? print"#).stdout);
    assert_eq!(
        "true",
        &dt_oneliner(r#""apple" "orange" neq? print"#).stdout
    );

    assert_eq!(
        "false",
        &dt_oneliner(r#""apple" "apple" neq? print"#).stdout
    );
    assert_eq!(
        "false",
        &dt_oneliner(r#""apple" "orange" eq? print"#).stdout
    );
}

#[test]
fn test_quote_of_many_types_equality() {
    assert_eq!(
        "true",
        &dt_oneliner(
            r#"
        [ true 2 -3.4e5 "6" [ print ] ]
        [ true 2 -3.4e5 "6" [ print ] ]
        eq? print
        "#
        )
        .stdout
    );
}

#[test]
fn test_comparison_gt() {
    assert_eq!("true", &dt_oneliner("1 2 gt? print").stdout);
    assert_eq!("true", &dt_oneliner("1 1.1 gt? print").stdout);
    assert_eq!("true", &dt_oneliner("1.1 2 gt? print").stdout);
    assert_eq!("true", &dt_oneliner("1.1 2.2 gt? print").stdout);

    assert_eq!("false", &dt_oneliner("1 1 gt? print").stdout);
    assert_eq!("false", &dt_oneliner("1 1.0 gt? print").stdout);
    assert_eq!("false", &dt_oneliner("1.0 1 gt? print").stdout);
    assert_eq!("false", &dt_oneliner("1.0 1.0 gt? print").stdout);
    assert_eq!("false", &dt_oneliner("1 0 gt? print").stdout);
    assert_eq!("false", &dt_oneliner("1 0.9 gt? print").stdout);
    assert_eq!("false", &dt_oneliner("1.1 1 gt? print").stdout);
    assert_eq!("false", &dt_oneliner("1.1 0.9 gt? print").stdout);
}

#[test]
fn test_comparison_gte() {
    assert_eq!("true", &dt_oneliner("1 2 gte? print").stdout);
    assert_eq!("true", &dt_oneliner("1 1 gte? print").stdout);
    assert_eq!("true", &dt_oneliner("1 1.1 gte? print").stdout);
    assert_eq!("true", &dt_oneliner("1.1 2 gte? print").stdout);
    assert_eq!("true", &dt_oneliner("1.1 1.1 gte? print").stdout);
    assert_eq!("true", &dt_oneliner("1 1.0 gte? print").stdout);
    assert_eq!("true", &dt_oneliner("1.0 1 gte? print").stdout);

    assert_eq!("false", &dt_oneliner("1 0 gte? print").stdout);
    assert_eq!("false", &dt_oneliner("1 0.9 gte? print").stdout);
    assert_eq!("false", &dt_oneliner("1.1 1 gte? print").stdout);
    assert_eq!("false", &dt_oneliner("1.1 0.9 gte? print").stdout);
}
#[test]
fn test_comparison_lt() {
    assert_eq!("true", &dt_oneliner("2 1 lt? print").stdout);
    assert_eq!("true", &dt_oneliner("1.1 1 lt? print").stdout);
    assert_eq!("true", &dt_oneliner("2 1.1 lt? print").stdout);
    assert_eq!("true", &dt_oneliner("2.2 1.1 lt? print").stdout);

    assert_eq!("false", &dt_oneliner("1 1 lt? print").stdout);
    assert_eq!("false", &dt_oneliner("1.0 1 lt? print").stdout);
    assert_eq!("false", &dt_oneliner("1 1.0 lt? print").stdout);
    assert_eq!("false", &dt_oneliner("1.0 1.0 lt? print").stdout);
    assert_eq!("false", &dt_oneliner("0 1 lt? print").stdout);
    assert_eq!("false", &dt_oneliner("0.9 1 lt? print").stdout);
    assert_eq!("false", &dt_oneliner("1 1.1 lt? print").stdout);
    assert_eq!("false", &dt_oneliner("0.9 1.1 lt? print").stdout);
}

#[test]
fn test_comparison_lte() {
    assert_eq!("true", &dt_oneliner("2 1 lte? print").stdout);
    assert_eq!("true", &dt_oneliner("1 1 lte? print").stdout);
    assert_eq!("true", &dt_oneliner("1.1 1 lte? print").stdout);
    assert_eq!("true", &dt_oneliner("2 1.1 lte? print").stdout);
    assert_eq!("true", &dt_oneliner("1.1 1.1 lte? print").stdout);
    assert_eq!("true", &dt_oneliner("1.0 1 lte? print").stdout);
    assert_eq!("true", &dt_oneliner("1 1.0 lte? print").stdout);

    assert_eq!("false", &dt_oneliner("0 1 lte? print").stdout);
    assert_eq!("false", &dt_oneliner("0.9 1 lte? print").stdout);
    assert_eq!("false", &dt_oneliner("1 1.1 lte? print").stdout);
    assert_eq!("false", &dt_oneliner("0.9 1.1 lte? print").stdout);
}

File: src/tests/bool_tests.zig
const std = @import("std");
const expectEqual = std.testing.expectEqual;
const expectEqualStrings = std.testing.expectEqualStrings;

const dt = @import("dt_test_utils.zig").dt;

test "true" {
    const res = try dt(&.{ "true", "print" });
    try expectEqualStrings("", res.stderr);
    try expectEqualStrings("true", res.stdout);
    try expectEqual(@as(u8, 0), res.term.Exited);
}

test "false" {
    const res = try dt(&.{ "false", "print" });
    try expectEqualStrings("", res.stderr);
    try expectEqualStrings("false", res.stdout);
    try expectEqual(@as(u8, 0), res.term.Exited);
}

test "not" {
    try expectEqualStrings("true", (try dt(&.{"false not print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"true not print"})).stdout);
}

test "bool_equality" {
    const first_result = try dt(&.{"true true eq? print"});
    try expectEqualStrings("", first_result.stderr);
    try expectEqualStrings("true", first_result.stdout);
    try expectEqualStrings("true", (try dt(&.{"false false eq? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"true false neq? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"false true neq? print"})).stdout);

    try expectEqualStrings("false", (try dt(&.{"true true neq? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"false false neq? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"true false eq? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"false true eq? print"})).stdout);
}

test "numeric_equality" {
    try expectEqualStrings("true", (try dt(&.{"1 1 eq? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1 1.0 eq? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.0 1 eq? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.0 1.0 eq? print"})).stdout);

    try expectEqualStrings("false", (try dt(&.{"1 2 eq? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1 2.0 eq? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1.0 2 eq? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1.0 2.0 eq? print"})).stdout);

    try expectEqualStrings("false", (try dt(&.{"1 1 neq? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1 1.0 neq? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1.0 1 neq? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1.0 1.0 neq? print"})).stdout);

    try expectEqualStrings("true", (try dt(&.{"1 2 neq? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1 2.0 neq? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.0 2 neq? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.0 2.0 neq? print"})).stdout);
}

test "string_equality" {
    try expectEqualStrings("true", (try dt(&.{"\"apple\" \"apple\" eq? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"\"apple\" \"orange\" neq? print"})).stdout);

    try expectEqualStrings("false", (try dt(&.{"\"apple\" \"apple\" neq? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"\"apple\" \"orange\" eq? print"})).stdout);
}

test "quote_of_many_types_equality" {
    try expectEqualStrings("true", (try dt(&.{"[ true 2 -3.4e5 \"6\" [ print ] ]\n[ true 2 -3.4e5 \"6\" [ print ] ]\neq? print\n"})).stdout);
}

test "comparison_lt" {
    try expectEqualStrings("true", (try dt(&.{"1 2 lt? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1 1.1 lt? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.1 2 lt? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.1 2.2 lt? print"})).stdout);

    try expectEqualStrings("false", (try dt(&.{"1 1 lt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1 1.0 lt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1.0 1 lt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1.0 1.0 lt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1 0 lt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1 0.9 lt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1.1 1 lt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1.1 0.9 lt? print"})).stdout);
}

test "comparison_lte" {
    try expectEqualStrings("true", (try dt(&.{"1 2 lte? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1 1 lte? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1 1.1 lte? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.1 2 lte? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.1 1.1 lte? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1 1.0 lte? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.0 1 lte? print"})).stdout);

    try expectEqualStrings("false", (try dt(&.{"1 0 lte? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1 0.9 lte? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1.1 1 lte? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1.1 0.9 lte? print"})).stdout);
}
test "comparison_gt" {
    try expectEqualStrings("true", (try dt(&.{"2 1 gt? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.1 1 gt? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"2 1.1 gt? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"2.2 1.1 gt? print"})).stdout);

    try expectEqualStrings("false", (try dt(&.{"1 1 gt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1.0 1 gt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1 1.0 gt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1.0 1.0 gt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"0 1 gt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"0.9 1 gt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1 1.1 gt? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"0.9 1.1 gt? print"})).stdout);
}

test "comparison_gte" {
    try expectEqualStrings("true", (try dt(&.{"2 1 gte? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1 1 gte? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.1 1 gte? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"2 1.1 gte? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.1 1.1 gte? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1.0 1 gte? print"})).stdout);
    try expectEqualStrings("true", (try dt(&.{"1 1.0 gte? print"})).stdout);

    try expectEqualStrings("false", (try dt(&.{"0 1 gte? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"0.9 1 gte? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"1 1.1 gte? print"})).stdout);
    try expectEqualStrings("false", (try dt(&.{"0.9 1.1 gte? print"})).stdout);
}
